cli/popper/parser.py

```diff
@@ -367,6 +380,96 @@ def _traverse(entrypoint, reachable, actions):
+                                variable in the workflow.
+
+        Returns:
+            None
+
+        """
+
+        substitution_dict = dict()
+
+        for args in substitutions:
+            item = args.split('=', 1)
+            if len(item) < 2:
+                raise Exception("Excepting '=' as seperator")
+            substitution_dict['$'+item[0]] = item[1]
+
+        for keys in substitution_dict:
+            if(not bool(re.match(r"\$_[A-Z0-9]+", keys))):
+                log.fail("Substitution variable '{}' doesn't "
+                         "satify required format ".format(keys))
+
+        used = {}
+
+        for wf_name, wf_block in self.action.items():
+
+            attr = wf_block.get('needs', [])
+            for i in range(len(attr)):
+                for k, v in substitution_dict.items():
+                    if k in attr[i]:
+                        used[k] = 1
+                        attr[i] = attr[i].replace(k, v)
+
+            attr = wf_block.get('uses', '')
+            for k, v in substitution_dict.items():
+                if k in attr:
+                    used[k] = 1
+                    wf_block['uses'] = attr.replace(k, v)
+
+            attr = wf_block.get('args', [])
+            for i in range(len(attr)):
+                for k, v in substitution_dict.items():
+                    if k in attr[i]:
+                        used[k] = 1
+                        attr[i] = attr[i].replace(k, v)
+
+            attr = wf_block.get('runs', [])
+            for i in range(len(attr)):
+                for k, v in substitution_dict.items():
+                    if k in attr[i]:
+                        used[k] = 1
+                        attr[i] = attr[i].replace(k, v)
+
+            attr = wf_block.get('secrets', [])
+            for i in range(len(attr)):
+                for k, v in substitution_dict.items():
+                    if k in attr[i]:
+                        used[k] = 1
+                        attr[i] = attr[i].replace(k, v)
+
+            attr = wf_block.get('env', {})
+            temp_dict = {}
+            for key in attr.keys():
+                check_replacement = False
+                for k, v in substitution_dict.items():
+                    if k in key:
+                        used[k] = 1
+                        temp_dict[v] = attr[key]
+                        check_replacement = True
+
+                if(check_replacement is False):
+                    temp_dict[key] = attr[key]
+
+            for key, value in temp_dict.items():
+                for k, v in substitution_dict.items():
+                    if k in value:
+                        used[k] = 1
+                        temp_dict[key] = v
+
+            if(len(temp_dict) != 0):
+                wf_block['env'] = temp_dict
+
+        if not allow_loose and len(substitution_dict) != len(used):
+            log.fail("All the substitutions passed are not used !")
```

Comment:
maybe `"Not all given substitutions are used in the workflow file"`

Reply:
Thanks ivo, for reviewing the pr, Fixed this part.